### PR TITLE
SALTO-1934: fetching security levels

### PR DESCRIPTION
**Pull request by alonstern**
Added security levels to security schemes

---

- Added `dataField` to recurse into to prevent when running the recurse into of project into security scheme to generate other standalone values for security scheme levels
- Fixed a bug in adapter-utils transformValues, that we would run `_.isEmpty` on numbers and boolean which would return true and we would assume the value is empty 

---
_Release Notes_: 
_Jira Adapter_:
- Added security levels to security schemes

---
_User Notifications_: 
_Jira Adapter_:
- After the next fetch, security scheme levels instances will be added to the workspace

### CODE REVIEW

**Review by netama — COMMENTED** [review]
partial review, will continue later - but I think it will be better to split into separate PRs :)

File: packages/adapter-components/src/elements/swagger/instance_elements.ts [hunk 1]
```diff
@@ -144,7 +144,7 @@ const getListDeepInnerType = async (
  */
 const normalizeElementValues = (instance: InstanceElement): Promise<InstanceElement> => {
   const transformAdditionalProps: TransformFunc = async ({ value, field, path }) => {
-    if (Array.isArray(value)) {
+    if (!_.isPlainObject(value)) {
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/instance_elements.ts:147]: We treat value as an object in the rest of the function so every other case is not valid. e.g. if value is a string, running `pickBy` with any condition function will convert it from e.g., `"abc"` to ```{
   0: 'a',
   1: 'b',
   2: 'c',
}```

File: packages/adapter-components/src/elements/swagger/instance_elements.ts [hunk 2]
```diff
@@ -370,7 +370,7 @@ const getEntriesForType = async (
 
   const getExtraFieldValues = (
     entry: Values
-  ): Promise<[string, Values | Values[]][]> => Promise.all(
+  ): Promise<[string, Value | Value[]][]> => Promise.all(
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/instance_elements.ts:373]: I don't think so, because the `dataField` the return value here does not have to be an object so it really could be anything

File: packages/adapter-components/src/elements/swagger/instance_elements.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (netama) [reply packages/adapter-components/src/elements/swagger/instance_elements.ts:147]: but if the data field is a primitive then we can't create an element from it, I think we might be missing some safeties here (or maybe I'm missing the exact scenario) - might be easier to discuss IRL :)

**Review by netama — COMMENTED** [review]
LGTM for the logic (except for the inconsistency between swagger and ducktype) - as discussed IRL, I am not 100% sure about the new parameter name. I would either go with a different name to avoid confusion, or (better if possible) use the `fullValue` serialization like we do in zendesk, and then we can keep the full objects (and I think in that case we also won't need the `dataField`).
if you want to keep the parameter, will be best if @ori-moisis can review as well as he added the `RecurseInto` feature.

let's also cover the new jira elements with some test (maybe even e2e)?

File: packages/adapter-components/src/config/request.ts [hunk 3]
```diff
@@ -55,6 +55,7 @@ type RecurseIntoConfig = {
   isSingle?: boolean
   context: RecurseIntoContext[]
   conditions?: RecurseIntoCondition[]
+  dataField?: string
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/config/request.ts:58]: 2. `isSingle` their is also only supported on swagger so I think we need to separate the types, can we open a ticket and do it as a different task?

**Comment by alonstern** [comment]
> LGTM for the logic (except for the inconsistency between swagger and ducktype) - as discussed IRL, I am not 100% sure about the new parameter name. I would either go with a different name to avoid confusion, or (better if possible) use the `fullValue` serialization like we do in zendesk, and then we can keep the full objects (and I think in that case we also won't need the `dataField`). if you want to keep the parameter, will be best if @ori-moisis can review as well as he added the `RecurseInto` feature.
> 
> let's also cover the new jira elements with some test (maybe even e2e)?

I will add an e2e after I will add a deploy (I must be able to deploy the instance to the service to test its fetching)

File: packages/adapter-components/src/config/request.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (netama) [reply packages/adapter-components/src/config/request.ts]: re `isSingle` - it should be covered in both as well, and have some vague recollection on commenting on that somewhere. I guess it was missed when the recurse-into ducktype support was initially added. 
if you're planning to support it immediately afterwards it can be an another PR, but if it will go in the backlog then I think we should not allow it to be configured (and have both of these as swagger-only parameters / blocked explicitly by validation)

File: packages/jira-adapter/src/config.ts [hunk 4]
```diff
@@ -1467,6 +1544,9 @@ export const DEFAULT_API_DEFINITIONS: JiraApiConfig = {
         newName: 'WorkflowSchemes',
       },
     ],
+    additionalTypes: [
+      { typeName: 'ProjectSecurityScheme', cloneFrom: 'SecurityScheme' },
```
Comment (netama) [inline packages/jira-adapter/src/config.ts:1548]: maybe add a comment explaining why we are cloning this type?

File: packages/jira-adapter/src/config.ts [hunk 5]
```diff
@@ -1648,7 +1728,7 @@ export const configType = createMatchingObjectType<Partial<JiraConfig>>({
     apiDefinitions: { refType: apiDefinitionsType },
   },
   annotations: {
-    [CORE_ANNOTATIONS.DEFAULT]: _.omit(DEFAULT_CONFIG, ['apiDefinitions', 'client']),
+    [CORE_ANNOTATIONS.DEFAULT]: DEFAULT_CONFIG,
```
Comment (netama) [inline packages/jira-adapter/src/config.ts:1731]: should this be reverted?

File: packages/adapter-components/src/elements/swagger/instance_elements.ts [hunk 6]
```diff
@@ -144,8 +144,7 @@ const getListDeepInnerType = async (
  */
 const normalizeElementValues = (instance: InstanceElement): Promise<InstanceElement> => {
   const transformAdditionalProps: TransformFunc = async ({ value, field, path }) => {
-    if (Array.isArray(value)) {
-      // will handle in inner call
+    if (!_.isPlainObject(value)) {
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/instance_elements.ts:147]: is this change still needed?

**Review by netama — APPROVED** [review]
🎉

File: packages/adapter-components/src/elements/swagger/instance_elements.ts (hunk 6 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/instance_elements.ts:147]: I think it's not happening currently but it's still a bug that might happen in future flows
